Add support for git remote resolution using provider APIs with an API token

abayer opened the pull request:
# Changes

Closes https://github.com/tektoncd/resolution/issues/42

This adds support for API-based token authentication and fetching of remote resources without requiring a full clone of the repository to the `git` resolver, using https://github.com/jenkins-x/go-scm for interacting with provider APIs. It is fully backwards compatible with the existing `git` resolver behavior and syntax, with the new behavior being used when the `repo` parameter is passed to the `git` resolver instead of `url`.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Adds new behavior to the git remote resolver to fetch Tasks and Pipelines from GitHub, Gitlab, BitBucket via their APIs using API tokens.
```

abayer commented:
/assign @lbernick @vdemeester @jerop

lbernick commented:
dumb question, SCM = source control manager right? How does a user know when to use the git resolver vs this one?

vdemeester commented:
> > dumb question, SCM = source control manager right? How does a user know when to use the git resolver vs this one?
> 
> Correct. Maybe it should be renamed "git-api"? In retrospect, I wish the git resolver had been named something different, like "anonymous-git", because the most important difference is that the scm resolver authenticates (using an API key), so it can access private repos. It also fetches files via the provider's API, so it doesn't need to do a full clone, etc.

- We can still rename 👼🏼 (before doing the release)
- Is that resolver support more than just `git` ? (and is the library we use for this support more than git ?) Depending on the error, we may want to rename this one with `git` in the name or not. `git-provider` would make more sense than `git-api` for me (as.. it is configurable and "tied" to a git provider, or a scm provider)

abayer commented:
`git-provider` does feel more right to me. Can anyone think of a shorter name for the existing non-authenticated, full-clone provider than `anonymous-git`?

vdemeester commented:
I meant `git-provider` for this resolver, not the anonymous one

abayer commented:
So did I. =)

So here's what I'm saying:
* The existing `git` resolver gets renamed to `anonymous-git`, or hopefully a better name than that.
* This new resolver gets renamed to `git-provider`.

lbernick commented:
Could we just have one git resolver that does not require credentials, but uses them if they are present?

abayer commented:
> Could we just have one git resolver that does not require credentials, but uses them if they are present?

The problem/challenge there is that their mechanisms and params differ - that said, I guess we could have it use the API if SCM provider/creds/org/repo are provided, and anonymous/full clone if they aren't and a URL is provided.

lbernick commented:
> > Could we just have one git resolver that does not require credentials, but uses them if they are present?
> 
> The problem/challenge there is that their mechanisms and params differ - that said, I guess we could have it use the API if SCM provider/creds/org/repo are provided, and anonymous/full clone if they aren't and a URL is provided.

I think that would be ideal if it's workable

abayer commented:
Ok, first attempt at merging the functionality is up - lots of cleanup to do tomorrow, assuming it works.

abayer commented:
/hold cancel

Ok, I have cleaned up my work from yesterday, removed the separate `scm` resolver, and updated docs etc accordingly. This should be good to go.

vicpadilla commented:
Has this approach been considered for fetching a single file?: https://stackoverflow.com/questions/28375418/git-how-to-pull-a-single-file-from-a-server-repository-in-git
It would allow us to continue using the go-git library which is provider-agnostic.

The go-git library also seems to support authentication: https://stackoverflow.com/a/48895644/2570297

abayer commented:
@vicpadilla - That has not - but it also looks like it still needs a full clone of the repo, and seems to be about updating a single file in a clone, not just fetching a single file without cloning, sadly. Could you open an issue for the go-git authentication? Thanks!

abayer commented:
/retest

lbernick commented on config/resolvers/config-feature-flags.yaml:
```diff
@@ -38,3 +38,7 @@ data:
   # This is an experimental feature and thus should still be considered
   # an alpha feature.
   enable-cluster-resolver: "false"
+  # Setting this flag to "true" enables remote resolution of tasks and pipelines via SCM provider APIs.
+  # This is an experimental feature and thus should still be considered
+  # an alpha feature.
+  enable-scm-resolver: "false"
```
I think this can be removed

abayer replied:
Whoops, I missed a reference! Fixing.

lbernick commented on pkg/resolution/resolver/git/resolver.go:
```diff
@@ -30,45 +31,73 @@ import (
 	gitcfg "github.com/go-git/go-git/v5/config"
 	"github.com/go-git/go-git/v5/plumbing"
 	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/jenkins-x/go-scm/scm/factory"
```
curious why you're using jenkins-x/go-scm instead of https://github.com/drone/go-scm (the one linked to in the docs)?

abayer replied:
I link to the drone one? Lemme fix that, should be jenkins-x. That's the one I _know_ works, since I spent months working on it. =)

lbernick commented on test/git-resolver/gitea.yaml:
```diff
@@ -0,0 +1,853 @@
+# GENERATED BY RUNNING `helm template -n default gitea gitea-charts/gitea > gitea.yaml`
```
what's this file for?

abayer replied:
Spinning up gitea in the cluster for the git API resolution e2e tests. I'll add a comment.

abayer commented on pkg/apis/config/resolver/feature_flags.go:
```diff
@@ -33,6 +33,8 @@ const (
 	// DefaultEnableClusterResolver is the default value for "enable-cluster-resolver".
 	DefaultEnableClusterResolver = false
+	// DefaultEnableSCMResolver is the default value for "enable-scm-resolver".
+	DefaultEnableSCMResolver = false
```
oops!

abayer commented on pkg/resolution/resolver/git/resolver.go:
```diff
@@ -79,117 +108,137 @@ func (r *Resolver) ValidateParams(ctx context.Context, params map[string]string)
+		return err
 	}
-
-	// TODO(sbwsg): validate repo url is well-formed, git:// or https://
-	// TODO(sbwsg): validate pathInRepo is valid relative pathInRepo
```
That was a mistake, putting it back.

abayer replied:
They'll be at the bottom of `populateDefaultParams` now.

lbernick commented on pkg/resolution/resolver/framework/testing/featureflag.go:
```diff
@@ -43,6 +43,11 @@ func ContextWithClusterResolverEnabled(ctx context.Context) context.Context {
 	return contextWithResolverEnabled(ctx, "enable-cluster-resolver")
 }
 
+// ContextWithSCMResolverEnabled returns a context containing a Config with the enable-scm-resolver feature flag enabled.
+func ContextWithSCMResolverEnabled(ctx context.Context) context.Context {
+	return contextWithResolverEnabled(ctx, "enable-scm-resolver")
+}
+
```
this can be removed too

abayer replied:
hahaha, how do I keep missing these? =)

lbernick reviewed (approved):
thanks Andrew-- approving in the sense of "this looks reasonable and has tests" but not in the sense of "I'm able to approve every line for correctness". maybe one change that would help would be splitting the gitea server creation into a separate PR

abayer commented:
fwiw, I'm planning to move the gitea setup stuff to plumbing after this has merged, as well as making it a bit more generic. There are other places that could take advantage of having tooling for setting up gitea, like https://github.com/tektoncd/plumbing/pull/1185, for example. So I think it's fine to rubber-stamp the gitea stuff for now and give it a more thorough review when I move it to plumbing.

chuangw6 commented on pkg/resolution/resolver/git/resolver.go:
```diff
@@ -217,25 +266,155 @@ func (r *Resolver) isDisabled(ctx context.Context) bool {
-	return map[string]string{
-		AnnotationKeyRevision:                     r.Revision,
-		resolutioncommon.AnnotationKeyContentType: YAMLContentType,
+func (r *resolvedGitResource) Annotations() map[string]string {
+	m := map[string]string{
+		internal.BuiltInAnnotation(AnnotationKeyRevision):                     r.Revision,
+		internal.BuiltInAnnotation(AnnotationKeyPath):                         r.Path,
+		internal.BuiltInAnnotation(resolutioncommon.AnnotationKeyContentType): yamlContentType,
+	}
+
+	if r.Org != "" {
+		m[internal.BuiltInAnnotation(AnnotationKeyOrg)] = r.Org
 	}
+	if r.Repo != "" {
+		m[internal.BuiltInAnnotation(AnnotationKeyRepo)] = r.Repo
+	}
+	if r.URL != "" {
+		m[internal.BuiltInAnnotation(AnnotationKeyURL)] = r.URL
+	}
+
+	return m
+}
```
Is it possible to explicitly set the annotation keys to be namespaced directly in `annotations.go` rather than doing it at time of writing the annotations i.e. `AnnotationKeyPath=internal.BuiltInAnnotation("path")`? Doing it here implicitly might mislead that the annotation key is just original `AnnotationKeyPath` rather than the name-spaced.

abayer replied:
Yeah, that could be done, I just got tired of copy-pasting `resolution.tekton.dev/` everywhere. =)

abayer replied:
...and I'm gonna do it. First a PR to update all the existing resolvers other than `git` (since that'll be dealt with in this PR), then updating this PR.

abayer replied:
Oh, wait, I added it in this PR in the first place! 🤦

abayer replied:
aaaand done.

vdemeester commented on config/resolvers/git-resolver-config.yaml:
```diff
@@ -22,9 +22,22 @@ metadata:
   fetch-timeout: "1m"
-  # The git url to fetch the remote resource from.
+  # The git url to fetch the remote resource from when using anonymous cloning.
   default-url: "https://github.com/tektoncd/catalog.git"
```
I think I missed this earlier 🤔 Why do we default to the catalog repo ?

abayer replied:
I have no idea, but it was already that way when I copied it over from resolution.git. =)

vdemeester commented on docs/git-resolver.md:
```diff
@@ -6,11 +6,13 @@ This Resolver responds to type `git`.
-| `pathInRepo` | Where to find the file in the repo.                                          | `/task/golang-build/0.3/golang-build.yaml`                  |
+| Param Name   | Description                                                                                                       | Example Value                                               |
+|--------------|-------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------|
+| `url`        | URL of the repo to fetch and clone anonymously. Either `url` or `org` and `repo` must be specified, but not both. | `https://github.com/tektoncd/catalog.git`                   |
+| `repo`       | The repository to find the resource in. Either `url` or `org` and `repo` must be specified, but not both.         | `pipeline`, `test-infra`                                    |
+| `org`        | The organization to find the repository in. Default can be set in [configuration](#configuration).                | `tektoncd`, `kubernetes`                                    |
```
It's `url` or `repo`+`org` right ? Not sure how to make it clear in the parameter list but we should try 👼🏼

abayer replied:
Yeah, the wording could be improved, but I haven't come up with better yet.

vdemeester reviewed (commented):
On question : there would have to be one resolver per provider, am I right ? If on my cluster I want to be able to fetch task/pipeline from both `github` and my internal `gitlab` (or even gitlab.com), I would need to run 2 resolvers with different names ? or is it even possible ? (I would love it if it was 👼🏼 )

abayer commented:
> On question : there would have to be one resolver per provider, am I right ? If on my cluster I want to be able to fetch task/pipeline from both `github` and my internal `gitlab` (or even gitlab.com), I would need to run 2 resolvers with different names ? or is it even possible ? (I would love it if it was 👼🏼 )

As of now, yes, we can only support one provider. I need to think more about how we can support multiple providers/credentials/etc, but I think that can come later.

abayer commented on pkg/resolution/resolver/git/resolver.go:
```diff
@@ -79,117 +107,137 @@ func (r *Resolver) ValidateParams(ctx context.Context, params map[string]string)
-		} else {
-			return nil, fmt.Errorf("default Git Revision was not set during installation of the git resolver")
+
+	if params[urlParam] != "" {
```
👍

vdemeester commented:
/lgtm
We can figure out the multiple providers later on as this is still alpha 👼🏼
